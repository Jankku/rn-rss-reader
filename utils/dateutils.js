import { DateTime } from 'luxon';

/**
 * For articles released today return time and date otherwise.
 * @param {string} pubDate - RFC2822 date
 * @returns {string} Formatted date
 */
export function RFC2822ToTimeOrDate(pubDate) {
  const dateObject = DateTime.fromRFC2822(pubDate);
  const formattedDate =
    dateObject.toRelativeCalendar() === 'today'
      ? dateObject.toLocaleString(DateTime.TIME_24_SIMPLE)
      : dateObject.toLocaleString(DateTime.DATE_SHORT);
  return formattedDate;
}

/**
 * @param {string} pubDate - RFC2822 date
 * @returns {string} Formatted date in DateTime format
 */
export function RFC2822ToDateTime(pubDate) {
  const dateObject = DateTime.fromRFC2822(pubDate);
  const formattedDate = dateObject.toLocaleString(DateTime.DATETIME_SHORT);
  return formattedDate;
}
