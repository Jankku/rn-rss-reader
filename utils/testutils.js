import { render } from '@testing-library/react-native';
import { ThemeContextProvider } from '../context/ThemeContext';
import { LocationContextProvider } from '../context/LocationContext';
import { RegionContextProvider } from '../context/RegionContext';

const AllProviders = ({ children }) => {
  return (
    <LocationContextProvider>
      <RegionContextProvider>
        <ThemeContextProvider>{children}</ThemeContextProvider>
      </RegionContextProvider>
    </LocationContextProvider>
  );
};

const customRender = (ui, options) => render(ui, { wrapper: AllProviders, ...options });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export { customRender as render };
