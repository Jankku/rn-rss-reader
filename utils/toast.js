import Toast from 'react-native-root-toast';

const showToast = (message) =>
  Toast.show(message, {
    position: 100,
    opacity: 0.9,
    shadow: false,
  });

export default showToast;
