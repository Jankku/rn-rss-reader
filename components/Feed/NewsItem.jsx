import { useTheme } from '@react-navigation/native';
import { Dimensions, Image, Pressable, Text, View } from 'react-native';
import { stringOrPlaceholder } from '../../utils/stringutils';

function NewsItem({ title, description, imageUrl, onPress }) {
  const { colors, dark } = useTheme();

  return (
    <Pressable onPress={onPress}>
      <View
        style={{
          flex: 1,
          flexDirection: 'row',
          paddingHorizontal: 12,
          paddingVertical: 16,
          width: Dimensions.get('window').width,
          backgroundColor: colors.background,
        }}
      >
        <View style={{ flex: 1 }}>
          <Text style={{ color: colors.text, fontWeight: '700' }}>{stringOrPlaceholder(title)}</Text>
          <Text style={{ color: colors.text }}>{stringOrPlaceholder(description)}</Text>
        </View>
        <View style={{ paddingLeft: 4 }}>
          <Image
            resizeMode="cover"
            style={{
              backgroundColor: `rgba(${dark ? '255, 255, 255' : '0, 0, 0'}, 0.1)`,
              width: 100,
              height: 100,
              marginTop: 6,
              marginLeft: 4,
              borderRadius: 8,
            }}
            source={{ uri: imageUrl }}
          />
        </View>
      </View>
    </Pressable>
  );
}

export default NewsItem;
