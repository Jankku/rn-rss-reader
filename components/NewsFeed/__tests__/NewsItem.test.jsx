import NewsItem from '../NewsItem';
import { render, fireEvent, waitFor } from '../../../utils/testutils';
import { Settings } from 'luxon';

describe('Test NewsItem', () => {
  beforeAll(() => {
    Settings.defaultZone = 'UTC+3';
    Settings.defaultLocale = 'en-FI';
    Settings.now = () => new Date('Wed, 06 Apr 2022 20:00:00 +0300').valueOf();
  });

  afterAll(() => {
    Settings.defaultZone = 'local';
    Settings.defaultLocale = DateTime.now().resolvedLocaleOptions().locale;
  });

  test('Should match snapshot', () => {
    const { toJSON } = render(
      <NewsItem
        title={'title'}
        description={'description'}
        imageUrl={'imageUrl'}
        pubDate={'Wed, 06 Apr 2022 18:00:00 +0300'}
        onPress={() => {}}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  test('Should call onPress', () => {
    const onPress = jest.fn();
    const { getByText } = render(
      <NewsItem
        title={'title'}
        description={'description'}
        imageUrl={'imageUrl'}
        pubDate={'Wed, 06 Apr 2022 18:00:00 +0300'}
        onPress={onPress}
      />
    );
    const item = getByText('title');

    fireEvent.press(item);
    expect(onPress).toHaveBeenCalledTimes(1);
  });

  test('Should use placeholder values', () => {
    const { getAllByText } = render(
      <NewsItem title={''} description={''} imageUrl={'imageUrl'} pubDate={''} onPress={() => {}} />
    );
    const textElements = getAllByText('-');
    waitFor(() => expect(textElements).toHaveLength(3));
  });
});
