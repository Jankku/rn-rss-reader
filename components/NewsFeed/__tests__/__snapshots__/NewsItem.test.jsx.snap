// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test NewsItem Should match snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#f7f7f7",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "paddingHorizontal": 12,
        "paddingVertical": 16,
        "width": 750,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#101010",
            "fontWeight": "700",
            "paddingBottom": 4,
          }
        }
      >
        title
      </Text>
      <Text
        style={
          Object {
            "color": "#101010",
            "paddingBottom": 6,
          }
        }
      >
        description
      </Text>
      <Text
        style={
          Object {
            "color": "#101010",
            "fontSize": 12,
          }
        }
      >
        18.00
      </Text>
    </View>
    <View
      style={
        Object {
          "paddingLeft": 4,
        }
      }
    >
      <Image
        resizeMode="cover"
        source={
          Object {
            "uri": "imageUrl",
          }
        }
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.1)",
            "borderRadius": 8,
            "height": 100,
            "marginLeft": 4,
            "marginTop": 6,
            "width": 100,
          }
        }
      />
    </View>
  </View>
</View>
`;
